src/beams/calculate_QP_deformation.hpp

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <Kokkos_Core.hpp>
+
+#include "src/math/quaternion_operations.hpp"
+#include "src/types.hpp"
+
+namespace openturbine {
+
+struct CalculateQPDeformation {
+    Kokkos::View<size_t*>::const_type qp_num_qps_per_element;
+    Kokkos::View<double** [3]>::const_type qp_x0_;
+    Kokkos::View<double** [4]>::const_type qp_r_;
+    Kokkos::View<double** [7]>::const_type qp_x_;
+    Kokkos::View<double** [3]> qp_deformation_;
+
+    KOKKOS_FUNCTION void operator()(const size_t i_elem, const size_t i_qp) const {
```

Comment:
I've been doing these types of loops using `TeamPolicy` and hierarchical parallelism.  It gets similar or better performance, lets us write unit tests based on just the per-element action, and avoids the need for the early exit condition.

This doesn't seem to be a performance critical loop, so up to you if you want to change it.

Reply:
I think it makes sense to do the hierarchical parallelism, but I didn't immediately know how to do it in this context and I needed to make sure this was working. Since this is currently only called from `vtkouput` and not in the main loop I'll add a note that this could be improved later. Or you're welcome to add a commit for it.

Reply:
I'll make a note to circle back to this one in a follow up PR - I have a bit of state built up in my current workspace and don't want to hold you up.